### PR TITLE
IFrame API improvement

### PR DESCRIPTION
**Pull request by hristoterezov**

### CODE REVIEW

File: modules/API/external/external_api.js [hunk 1]
```diff
@@ -1,80 +1,66 @@
...
+const events = {
+    'displayNameChange': 'display-name-change',
+    'incomingMessage': 'incoming-message',
+    'outgoingMessage': 'outgoing-message',
+    'participantJoined': 'participant-joined',
+    'participantLeft': 'participant-left',
+    'readyToClose': 'video-ready-to-close',
+    'videoConferenceJoined': 'video-conference-joined',
+    'videoConferenceLeft': 'video-conference-left'
+};
```
Comment (virtuacoplenny) [inline modules/API/external/external_api.js]: While your gonna be all over the blame anyway, maybe it's time to get rid of the quotations around the keys (if that's the preferred style).

File: modules/API/external/external_api.js [hunk 2]
```diff
@@ -84,323 +70,342 @@ function changeParticipantNumber(APIInstance, number) {
...
+     * @private
+     */
+    _createIFrame(height, width) {
+        this.iframeHolder
+            = this.parentNode.appendChild(document.createElement('div'));
+        this.iframeHolder.id = `jitsiConference${id}`;
+        this.iframeHolder.style.width = `${width}px`;
+        this.iframeHolder.style.height = `${height}px`;
+        this.frameName = `jitsiConferenceFrame${id}`;
+        this.frame = document.createElement('iframe');
```
Comment (virtuacoplenny) [inline modules/API/external/external_api.js]: Silly nit: adding a space here might be good to denote that you are now working on a different node.

File: modules/API/external/external_api.js [hunk 3]
```diff
@@ -84,323 +70,342 @@ function changeParticipantNumber(APIInstance, number) {
...
+     * options defined in config.js to be overridden.
+     * @param {Object} [interfaceConfigOverwrite] - Object containing
+     * configuration options defined in interface_config.js to be overridden.
+     * @param {boolean} [noSSL] - If the value is true https won't be used.
+     * @param {string} [jwt] - The JWT token if needed by jitsi-meet for
+     * authentication.
+     */
+    constructor(domain, // eslint-disable-line max-params
+        roomName = '', width = MIN_WIDTH, height = MIN_HEIGHT,
+        parentNode = document.body, configOverwrite = {},
```
Comment (virtuacoplenny) [inline modules/API/external/external_api.js]: Is there a reason why you changed the default behavior of appending to the last script tag if parentNode is not defined?

File: doc/api.md [hunk 4]
```diff
@@ -123,12 +123,13 @@ You can also execute multiple commands using the `executeCommands` method:
...
 api.executeCommands(commands)
 ```
-The `commands` parameter is an object with the names of the commands as keys and the arguments for the commands asvalues:
+The `commands` parameter is an object with the names of the commands as keys and the arguments for the commands as values:
 ```javascript
 api.executeCommands({displayName: ['nickname'], toggleAudio: []});
 ```
 
 You can add event listeners to the embedded Jitsi Meet using the `addEventListener` method.
+*NOTE: This method still exists but it is deprecated. JitsiMeetExternalAPI class extends [EventEmitter]. Use [EventEmitter] methods (`addListener` or `on`).*
```
Comment (saghul) [inline doc/api.md]: finish the line with `**`

File: doc/api.md [hunk 5]
```diff
@@ -198,6 +199,7 @@ changes. The listener will receive an object with the following structure:
 You can also add multiple event listeners by using `addEventListeners`.
 This method requires one argument of type Object. The object argument must
 have the names of the events as keys and the listeners of the events as values.
+*NOTE: This method still exists but it is deprecated. JitsiMeetExternalAPI class extends [EventEmitter]. Use [EventEmitter] methods.*
```
Comment (saghul) [inline doc/api.md]: ditto

File: doc/api.md [hunk 6]
```diff
@@ -216,12 +218,13 @@ api.addEventListeners({
 ```
 
 If you want to remove a listener you can use `removeEventListener` method with argument the name of the event.
-
+*NOTE: This method still exists but it is deprecated. JitsiMeetExternalAPI class extends [EventEmitter]. Use [EventEmitter] methods( `removeListener`).*
```
Comment (saghul) [inline doc/api.md]: ditto

File: doc/api.md [hunk 7]
```diff
@@ -216,12 +218,13 @@ api.addEventListeners({
...
 ```javascript
 api.removeEventListener("incomingMessage");
 ```
 
 If you want to remove more than one event you can use `removeEventListeners` method with an Array with the names of the events as an argument.
+*NOTE: This method still exists but it is deprecated. JitsiMeetExternalAPI class extends [EventEmitter]. Use [EventEmitter] methods.*
```
Comment (saghul) [inline doc/api.md]: ditto

File: modules/API/external/external_api.js (hunk 1 above)
Comment (saghul) [inline modules/API/external/external_api.js]: Yes! 👍

File: modules/API/external/external_api.js [hunk 8]
```diff
@@ -84,323 +70,342 @@ function changeParticipantNumber(APIInstance, number) {
...
+     * @param {Object} [configOverwrite] - Object containing configuration
+     * options defined in config.js to be overridden.
+     * @param {Object} [interfaceConfigOverwrite] - Object containing
+     * configuration options defined in interface_config.js to be overridden.
+     * @param {boolean} [noSSL] - If the value is true https won't be used.
+     * @param {string} [jwt] - The JWT token if needed by jitsi-meet for
+     * authentication.
+     */
+    constructor(domain, // eslint-disable-line max-params
+        roomName = '', width = MIN_WIDTH, height = MIN_HEIGHT,
```
Comment (saghul) [inline modules/API/external/external_api.js]: note that if the room name is not specified a random one is not generated, so this doesn't work as advertised

Comment (saghul) [inline modules/API/external/external_api.js]: can you line these up vertically? If will be easier to visually parse...

File: modules/API/external/external_api.js [hunk 9]
```diff
@@ -84,323 +70,342 @@ function changeParticipantNumber(APIInstance, number) {
...
+        parentNode = document.body, configOverwrite = {},
+        interfaceConfigOverwrite = {}, noSSL = false, jwt = undefined) {
+        super();
+        this.parentNode = parentNode;
+        this.url = generateURL(domain, {
```
Comment (saghul) [inline modules/API/external/external_api.js]: why not make the domain part of the options object since we're here?

File: modules/API/external/external_api.js [hunk 10]
```diff
@@ -84,323 +70,342 @@ function changeParticipantNumber(APIInstance, number) {
...
+        this.iframeHolder.style.height = `${height}px`;
+        this.frameName = `jitsiConferenceFrame${id}`;
+        this.frame = document.createElement('iframe');
+        this.frame.src = this.url;
+        this.frame.name = this.frameName;
+        this.frame.id = this.frameName;
+        this.frame.width = '100%';
+        this.frame.height = '100%';
+        this.frame.setAttribute('allowFullScreen', 'true');
+        this.frame = this.iframeHolder.appendChild(this.frame);
```
Comment (saghul) [inline modules/API/external/external_api.js]: since you're here, can you set `frame.style.border = 0;` ?

File: modules/API/external/external_api.js [hunk 11]
```diff
@@ -84,323 +70,342 @@ function changeParticipantNumber(APIInstance, number) {
...
+        super();
+        this.parentNode = parentNode;
+        this.url = generateURL(domain, {
+            configOverwrite,
+            interfaceConfigOverwrite,
+            jwt,
+            noSSL,
+            roomName
+        });
+        this._createIFrame(Math.max(height, MIN_HEIGHT),
```
Comment (saghul) [inline modules/API/external/external_api.js]: what if width is "100%" ?

File: modules/API/external/external_api.js (hunk 9 above)
Comment (hristoterezov) [inline modules/API/external/external_api.js]: Because it is mandatory.

File: modules/API/external/external_api.js (hunk 10 above)
Comment (hristoterezov) [inline modules/API/external/external_api.js]: I prefer to do it trough another PR. Not sure if there aren't any better options to do this.

File: modules/API/external/external_api.js (hunk 3 above)
Comment (hristoterezov) [inline modules/API/external/external_api.js]: Actually the previous implementation was getting the parent of the last script tag. I guess for the users of the API it will be either the same (the iframe will be attached to the body) either the iframe was created in the head tag which doesn't make sense. I think even the original intention was to use the body tag as parent of the iframe.

Do you have any better ideas? Do you have in mind any issues that will be caused by this change?

File: modules/API/external/external_api.js (hunk 11 above)
Comment (hristoterezov) [inline modules/API/external/external_api.js]: Yeah! This will be an issue :( Good catch!

But actually I don't think we have been handling this case even before... I would say this implementation is equivalent to the previous one!

Since we have plans to deprecate those arguments or remove them or something else I don't think it worth to fix this now! We can address this one, once we decide what to do with the arguments in different PR.

File: modules/API/external/external_api.js (hunk 3 above)
Comment (saghul) [inline modules/API/external/external_api.js]: I think using the body makes sense. Though I'd rather make it mandatory.

File: modules/API/external/external_api.js (hunk 9 above)
Comment (saghul) [inline modules/API/external/external_api.js]: I mean the internal options object. That's not exposed to the user, so it's totally up to us.

File: modules/API/external/external_api.js (hunk 10 above)
Comment (saghul) [inline modules/API/external/external_api.js]: Why? You're already departing from the previous stuff, like those 100% 2 lines above.

File: modules/API/external/external_api.js (hunk 3 above)
Comment (virtuacoplenny) [inline modules/API/external/external_api.js]: Nope, I don't have other ideas or issues I can think of. I just noticed most of this PR tried to avoid modifying existing behavior but this changed it a little.

File: modules/API/external/external_api.js (hunk 8 above)
Comment (hristoterezov) [inline modules/API/external/external_api.js]: I think this depends on the welcome page setting. If there is no welcome page random room name is generated. Or we have changed the implementation since the last time I checked it?

Do you mean I should update the comment?

File: modules/API/external/external_api.js (hunk 9 above)
Comment (hristoterezov) [inline modules/API/external/external_api.js]: But still, I like it better to list the mandatory arguments. In this case it is only one so it is not a problem.

File: modules/API/external/external_api.js (hunk 10 above)
Comment (hristoterezov) [inline modules/API/external/external_api.js]: Which lines? If you mean `this.frame.width = '100%'` this is old code!

I don't think this change fits anywhere in this commits and the PR. Also I'm not sure whether I'd implement it with `frame.style.border = 0;` (of course we can start with this and change it later). 

Also in this PR there aren't any visual changes.
